substrate/client/rpc/src/state/state_full.rs

```diff
@@ -219,38 +253,57 @@ where
-		let mut storage_diff : Vec<(StorageKey, Option<StorageData>)> = vec![];
-		for key in end_keys {
-			if !start_keys.contains(&key) {
-				// if a new key is present, add it to the diff
-				let new_storage = self.client.storage(end, &key).map_err(client_err)?;
-				storage_diff.push((key, new_storage))
-			} else {
-				// else if the value has changed, add it to diff
-				let start_storage_val = self.client.storage(start, &key).map_err(client_err)?;
-				let end_storage_val = self.client.storage(end, &key).map_err(client_err)?;
+		let start_keys : Vec<_> = self.client.storage_keys(start, None, None).map_err(client_err)?.collect();
+		let end_keys : Vec<_> = self.client.storage_keys(end, None, None).map_err(client_err)?.collect();
 
-				if start_storage_val != end_storage_val {
-					storage_diff.push((key, end_storage_val))
-				}
-			}
-		}
+		self.prepare_storage_diff(start,end, start_keys, end_keys)
+	}
 
-		// finally get any keys that have been removed between start_block and end_block
-		for key in start_keys {
-			if storage_diff.iter().position(|r| r.0 == key).is_none() {
-				storage_diff.push((key, None))
-			}
+	fn storage_diff_with_prefixes(
+		&self,
+		start : Block::Hash,
+		end : Block::Hash,
+		prefixes : Vec<StorageKey>
+	) -> std::result::Result<Vec<(StorageKey, Option<StorageData>)>, JsonRpseeError> {
+		let mut start_keys = vec![];
+		let mut end_keys = vec![];
+
+		// read all storage keys with given prefixes
+		for prefix in prefixes {
+			let start_keys_of_prefix = self.client.storage_keys(start, None, Some(&prefix)).map_err(client_err)?;
+			let end_keys_of_prefix = self.client.storage_keys(end, None, Some(&prefix)).map_err(client_err)?;
+			start_keys.extend(start_keys_of_prefix);
+			end_keys.extend(end_keys_of_prefix);
 		}
+		
+		self.prepare_storage_diff(start,end, start_keys, end_keys)
+	}
 
-		Ok(storage_diff)
+	fn storage_diff_without_prefixes(
+		&self,
+		start : Block::Hash,
+		end : Block::Hash,
+		prefixes : Vec<StorageKey>
+	) -> std::result::Result<Vec<(StorageKey, Option<StorageData>)>, JsonRpseeError> {
+		let start_keys = self.client.storage_keys(start, None, None).map_err(client_err)?;
+		let end_keys = self.client.storage_keys(end, None, None).map_err(client_err)?;
+		let mut start_keys_to_exclude = vec![];
+		let mut end_keys_to_exclude = vec![];
+
+		// read all storage keys with given prefixes
+		for prefix in prefixes {
+			let start_keys_of_prefix = self.client.storage_keys(start, None, Some(&prefix)).map_err(client_err)?;
+			let end_keys_of_prefix = self.client.storage_keys(end, None, Some(&prefix)).map_err(client_err)?;
+			start_keys_to_exclude.extend(start_keys_of_prefix);
+			end_keys_to_exclude.extend(end_keys_of_prefix);
+		}
 
+		let start_keys : Vec<_> = start_keys.filter(|x| !start_keys_to_exclude.contains(&x)).collect();
+		let end_keys : Vec<_> = end_keys.filter(|x| !end_keys_to_exclude.contains(&x)).collect();
+		
+		self.prepare_storage_diff(start,end, start_keys, end_keys)
 	}
```

Comment:
Two points for the current solution.

1) The code performs many reads of keys `self.client.storage_keys(...)`. Means many calls to the database. This will deteriorate the performance if many prefixes specified in the params of the methods.

Can we just read all the keys for a particular block and then filter by prefixes manually?

2) lets combine two functions `storage_diff_with_prefixes`/`storage_diff_without_prefixes` in one. 
It should have two optional params: prefixes to include, prefixes to exclude.
By doing this the method will provide more flexibility than these two separate methods.

For example:
- prefixes to include: provide a prefix of the pallet Z;
- prefix to exclude: provide the prefix of the storage collection Y of the pallet Z;
- as a result - method will return all storage elements of only the pallet Z, but without the storage collection Y of that pallet.
- You can achieve the same with `storage_diff_with_prefixes` only by directly specifying all storage collections in the list of prefixes. Which is not alway convenient.